core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -72,6 +95,58 @@ public FilteredQueryBuilder queryName(String queryName) {
         return this;
     }
 
+    /** Returns the name of this query. */
+    public String queryName() {
+        return queryName;
+    }
+
+    @Override
+    protected boolean doEquals(FilteredQueryBuilder other) {
+        return Objects.equals(queryBuilder, other.queryBuilder) &&
+                Objects.equals(filterBuilder, other.filterBuilder) &&
+                Objects.equals(boost, other.boost) &&
+                Objects.equals(queryName, other.queryName);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(queryBuilder, filterBuilder, boost, queryName);
+    }
+
+    @Override
+    public Query toQuery(QueryParseContext parseContext) throws QueryParsingException, IOException {
+        // parsed internally, but returned null during parsing...
+        if (queryBuilder == null) {
+            return null;
+        }
+
+        if (filterBuilder == null) {
+            return queryBuilder.toQuery(parseContext);
+        }
+
+        Query query = queryBuilder.toQuery(parseContext);
+        Query filter = filterBuilder.toQuery(parseContext);
+
+        if (Queries.isConstantMatchAllQuery(filter)) {
+            // this is an instance of match all filter, just execute the query
+            return queryBuilder.toQuery(parseContext);
+        }
+
+        // if its a match_all query, use constant_score
+        if (Queries.isConstantMatchAllQuery(query)) {
+            Query q = new ConstantScoreQuery(filter);
+            q.setBoost(boost);
+            return q;
+        }
+
+        BooleanQuery filteredQuery = Queries.filtered(queryBuilder.toQuery(parseContext), filter);
+        filteredQuery.setBoost(boost);
+        if (queryName != null) {
```

Comment:
I just realize that there might be a bug in the existing code already. We only seem to add the query to the named queries if it's a BooleanQuery, the other cases return early. Not sure, but we might want to change that.